### PR TITLE
fix slow things

### PR DESCRIPTION
**Pull request by Rratic**
so many slow factors here

### CODE REVIEW

**Comment by Rratic** [comment]
May anyone give a `review` ?
The repo has been stuck for a week!

**Comment by Azzaare** [comment]
@Rratic thank you for those changes, they are indeed improvements :)

As we are slowing taking care of the accumulated PR, we can merge those changes later, but they will need some fix due to conflicts introduced by each previous merged PR. I try to give priority in order of submission, so it can wait for PR #53 to #56 to be merged!

File: src/math/abs.jl
Comment (Ved-Mahajan) [inline src/math/abs.jl:59]: I guess `Number` can be changed to `Real` becuase comparing `Complex` does not make sense

File: src/math/abs.jl
Comment (Ved-Mahajan) [inline src/math/abs.jl:37]: I guess Number can be changed to Real becuase comparing Complex does not make sense

File: src/math/abs.jl
Comment (Ved-Mahajan) [inline src/math/abs.jl:18]: I guess Number can be changed to Real becuase comparing Complex does not make sense

File: src/math/average_mean.jl
Comment (Ved-Mahajan) [inline src/math/average_mean.jl:16]: I guess this should be better: `mean(nums::Vector)` to `mean(nums::Vector{Number})`

File: src/math/average_median.jl
Comment (ReneganRonin) [inline src/math/average_median.jl:16]: try `function median(nums::Vector{T}) where T <: Number` because `Any` is default in Vector

File: src/math/average_mean.jl
Comment (ReneganRonin) [reply src/math/average_mean.jl:16]: tried it. it should be `mean(nums::Vector{T}) where T <: Number`

**Comment by ReneganRonin** [comment]
i am not sure why codecov failed. probably a missed test?

**Comment by Rratic** [comment]
Well, personly, I think `CI` is enough. We ourselves know what we are doing. Tests are not necessary, `codecov` only adds cost . . .

**Comment by ReneganRonin** [comment]
🤔 I disagree that tests are unnecessary. i dont have much of an opinion regarding codecov since i dont use it.

**Comment by Rratic** [comment]
Well, you see, I test them myself, and every time we test, we test the whole thing, which takes much time and energy that we already know to be no use, as we already know they are correct.

**Comment by Ved-Mahajan** [comment]
> Well, you see, I test them myself, and every time we _test_, we test the whole thing on different OS, which takes much time and energy that we already know to be no use ,as we already know they are correct.(julia is portable and so do algorithms!)

well ummm you have tested it and you know it works and that how much of the code is actually being tested. I trust you and so I know that your tests run fine, but what about a random person looking at the code?
Codecov is kind of an assurance that the how much code is tested and also checks if the code is actually working as per the tests.
This is my opinion tough. @Azzaare is the best to talk about this.

**Comment by Azzaare** [comment]
Well, we're all entitled to have an opinion about how to organize collaborative code. That being said, here are my reasons to have CI and not decreasing coverage:
- Reviewers (such as myself) can have a much efficient look at the code
- Collaborators that modify an existing algorithm can see if they mess up things (and that cannot even be guaranteed with full coverage, but it still is more likely)
- Versions of Julia and other dependencies will change in the future. Maintainers here can pinpoint the places to update code thanks to CI. Once again, the more coverage you have, the more likely you are to not miss any compatibility issue

As for the portability of Julia ... By experience, I can tell you it is not always the case, in the sense that it is not straightforward to make 100% portable code. It is relatively easy on average though compared to other languages. Still, it doesn't cost anything to cover all possible hardware (coupled with stable and nightly versions of Julia).

Here, it is easy to reach full coverage:
- test case of `n = 1`
- use [`@test_throws`](https://docs.julialang.org/en/v1/stdlib/Test/#Test.@test_throws) for `n < 0`

Finally, we can still merge things without full coverage. Sometimes, it is just too hard to test everything. But the current situation is not like that.

**Comment by Rratic** [comment]
> but what about a random person looking at the code?

the description here says
> Open Source resource for learning Data Structures & Algorithms and their implementation in any Programming Language

In this case , users shall focus on the `implementation`, rather than waste time testing.

Fine,I will see which I didn't `test` ...

File: src/math/abs.jl
Comment (Azzaare) [reply src/math/abs.jl]: Unless there is a conflict with another `abs_min` method, there is no need to provide types

File: src/math/abs.jl
Comment (Azzaare) [reply src/math/abs.jl]: Unless there is a conflict with another `abs_max` method, there is no need to provide types

File: src/math/abs.jl
Comment (Azzaare) [reply src/math/abs.jl]: Unless there is a conflict with another `abs_val` method, there is no need to provide types

File: src/math/average_mean.jl
Comment (Azzaare) [reply src/math/average_mean.jl]: Also, unless there is a conflict with another `mean` method, there is no need to provide types

File: src/math/average_median.jl
Comment (Azzaare) [reply src/math/average_median.jl]: Also, unless there is a conflict with another median method, there is no need to provide types

**Comment by Azzaare** [comment]
@Rratic Thank you for the update

I would appreciate if you don't use force push, as it erases the work of others (here it was just rebasing on main, so it was easy to redo but ...)

Please use `git pull` then `git push`

**Review by Azzaare — APPROVED** [review]

**Pull request merged**
